src/services/obs/backend.rs

```diff
@@ -0,0 +1,350 @@
+            self.root = Some(root.to_string())
+        }
+
+        self
+    }
+
+    /// Set endpoint of this backend.
+    ///
+    /// Both huaweicloud default domain and user domain endpoints are allowed.
+    /// Please DO NOT add the bucket name to the endpoint.
+    ///
+    /// - `https://obs.cn-north-4.myhuaweicloud.com`
+    /// - `obs.cn-north-4.myhuaweicloud.com` (https by default)
+    /// - `http://127.0.0.1:9000`
+    pub fn endpoint(&mut self, endpoint: &str) -> &mut Self {
+        if !endpoint.is_empty() {
+            self.endpoint = Some(endpoint.trim_end_matches('/').to_string());
+        }
+
+        self
+    }
+
+    /// Set access_key_id of this backend.
+    /// - If it is set, we will take user's input first.
+    /// - If not, we will try to load it from environment.
+    pub fn access_key_id(&mut self, access_key_id: &str) -> &mut Self {
+        if !access_key_id.is_empty() {
+            self.access_key_id = Some(access_key_id.to_string());
+        }
+
+        self
+    }
+
+    /// Set secret_access_key of this backend.
+    /// - If it is set, we will take user's input first.
+    /// - If not, we will try to load it from environment.
+    pub fn secret_access_key(&mut self, secret_access_key: &str) -> &mut Self {
+        if !secret_access_key.is_empty() {
+            self.secret_access_key = Some(secret_access_key.to_string());
+        }
+
+        self
+    }
+
+    /// Set bucket of this backend.
+    /// The param is required.
+    pub fn bucket(&mut self, bucket: &str) -> &mut Self {
+        if !bucket.is_empty() {
+            self.bucket = Some(bucket.to_string());
+        }
+
+        self
+    }
+
+    /// Consume builder to build an OBS backend.
+    pub fn build(&mut self) -> Result<Backend> {
+        info!("backend build started: {:?}", &self);
+
+        let root = match &self.root {
+            // Use "/" as root if user not specified.
+            None => "/".to_string(),
+            Some(v) => {
+                let mut v = v
+                    .split('/')
+                    .filter(|v| !v.is_empty())
+                    .collect::<Vec<&str>>()
+                    .join("/");
+                if !v.starts_with('/') {
+                    v.insert(0, '/');
+                }
+                if !v.ends_with('/') {
+                    v.push('/')
+                }
+                v
+            }
+        };
+
+        info!("backend use root {}", root);
+
+        let bucket = match &self.bucket {
+            Some(bucket) => Ok(bucket.to_string()),
+            None => Err(other(BackendError::new(
+                HashMap::from([("bucket".to_string(), "".to_string())]),
+                anyhow!("bucket is empty"),
+            ))),
+        }?;
+        debug!("backend use bucket {}", &bucket);
+
+        let is_obs_default_endpoint = OBS_DEFAULT_ENDPOINT_PATTERN.is_match(&endpoint);
+
+        let endpoint = match &self.endpoint {
+            Some(endpoint) => {
```

Comment:
Parse endpoint into `url::URL` so that we don't need to `replace` or `trim_matches`.

Reply:
Use `http::uri::Uri`: https://docs.rs/http/latest/http/uri/struct.Uri.html

The same type used in our HTTP client.

Reply:
Do you mean parse endpoint to Url and store it in Backend?

```rust
pub struct Backend {
    client: HttpClient,
    root: String,
    endpoint: Url,
    signer: Arc<Signer>,
    bucket: String,
}
```

Reply:
No, just use it during `build`. We still store as a `string` in Backend because we will not check or change it anymore.

Reply:
got